### PR TITLE
preact-base-element: modernize statics

### PR DESCRIPTION
**Pull request by samouri**
**summary**
Context: split from https://github.com/ampproject/amphtml/pull/37322

Now that we are no longer using `Closure Compiler` for downleveling, we can take better advantage of esnext features like `static` functions and variables. We also no longer need to use brackets for property access, since `terser` only mangles private properties.

- This PR modernizes `PreactBaseElement`
- It also moves `wrapRefProperty` into PBE, since it requires usage of private properties 😱

### CODE REVIEW

File: src/preact/base-element.js [hunk 1]
```diff
@@ -131,6 +131,104 @@ export class PreactBaseElement extends BaseElement {
     return !this.usesLoading();
   }
 
+  /**
+   * Override to provide the Component definition.
+   *
+   * @protected {!PreactDef.FunctionalComponent}
```
Comment (rcebulko) [inline src/preact/base-element.js:137]: `!`

**Review by rcebulko — APPROVED** [review]
approval pending removing !

File: src/preact/base-element.js (hunk 1 above)
Comment (samouri) [reply src/preact/base-element.js:137]: There were so many to remove. You started a 🚋

File: src/preact/base-element.js [hunk 2]
```diff
@@ -39,23 +39,23 @@ import {
 
 import {installShadowStyle} from '../shadow-embed';
 
-/** @const {!MutationObserverInit} */
+/** @const {MutationObserverInit} */
```
Comment (rcebulko) [inline src/preact/base-element.js:42]: Does `@const` actually type?

File: src/preact/base-element.js [hunk 3]
```diff
@@ -64,13 +64,13 @@ const SHADOW_CONTAINER_ATTRS = dict({
...
 
-/** @const {!JsonObject<string, string>} */
+/** @const {JsonObject<string, string>} */
 const SERVICE_SLOT_ATTRS = dict({'name': SERVICE_SLOT_NAME});
 
 /** @const {string} */
 const RENDERED_ATTR = 'i-amphtml-rendered';
 
-/** @const {!JsonObject<string, string>} */
+/** @const {JsonObject<string, string>} */
```
Comment (rcebulko) [inline src/preact/base-element.js:73]: JsonObject isn't a generic

File: src/preact/base-element.js (hunk 2 above)
Comment (samouri) [reply src/preact/base-element.js:42]: This PR doesn't address type checking

File: src/preact/base-element.js (hunk 3 above)
Comment (samouri) [reply src/preact/base-element.js:73]: This PR doesn't address type checking.

Comment (rcebulko) [reply src/preact/base-element.js:73]: Oh I didn't realize this wasn't type-checked. Carry on then 👍🏻

**Review by rcebulko — APPROVED** [review]
Missed that this doesn't address type-checking.

File: src/preact/base-element.js [hunk 4]
```diff
@@ -837,129 +933,30 @@ export class PreactBaseElement extends BaseElement {
...
- * when elements within the shadow DOM gain focus, the focus is also applied
- * to the host element.
- */
-PreactBaseElement['delegatesFocus'] = false;
+  /**
+   * @param {Object} api
+   * @param {string} key
+   */
+  wrapRefProperty(api, key) {
+    // eslint-disable-next-line local/restrict-this-access
```
Comment (jridgewell) [inline src/preact/base-element.js:942]: Likely unneeded now that it's an instance method.

File: src/preact/base-element.js [hunk 5]
```diff
@@ -837,129 +933,30 @@ export class PreactBaseElement extends BaseElement {
...
- * this component.  A key feature of 'delegatesFocus' set to true is that
- * when elements within the shadow DOM gain focus, the focus is also applied
- * to the host element.
- */
-PreactBaseElement['delegatesFocus'] = false;
+  /**
+   * @param {Object} api
+   * @param {string} key
+   */
+  wrapRefProperty(api, key) {
```
Comment (jridgewell) [inline src/preact/base-element.js:941]: This should be made private.

**Review by jridgewell — APPROVED** [review]

File: src/preact/base-element.js (hunk 4 above)
Comment (samouri) [reply src/preact/base-element.js:942]: `this` access is banned. it wasn't necessary when calling `this.currentRef_` _but_ that doesn't work b/c of the `this` changes within scope of the defineProperty getter/setter.

File: src/preact/base-element.js (hunk 5 above)
Comment (samouri) [reply src/preact/base-element.js:941]: Done

File: src/preact/base-element.js (hunk 4 above)
Comment (samouri) [reply src/preact/base-element.js]: Ah, fixed with arrow func. Removed now.

Comment (jridgewell) [reply src/preact/base-element.js]: Use arrows instead of methods

**Pull request merged**